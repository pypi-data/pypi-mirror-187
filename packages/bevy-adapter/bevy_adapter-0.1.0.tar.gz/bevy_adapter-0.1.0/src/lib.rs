use bevy_reflect::serde::ReflectSerializer;
use bevy_reflect::{DynamicStruct, Reflect, TypeRegistry};
use bevy_transform::components::Transform;
use glam::{Quat, Vec3};
use pyo3::prelude::*;

/// <Vector (0, 1, 2)>
#[derive(FromPyObject)]
pub struct Vector(f32, f32, f32);

impl Vector {
    fn to_vec3(self) -> Vec3 {
        Vec3::new(self.0, self.1, self.2)
    }
}

/// <Quaternion (w=0, x=1, y=2, z=3)>
#[derive(FromPyObject)]
pub struct Quaternion(f32, f32, f32, f32);

impl Quaternion {
    fn to_quat(self) -> Quat {
        Quat::from_xyzw(self.3, self.2, self.1, self.0)
    }
}

/// Serializes transform into ron string.
#[pyfunction]
pub fn transform_ron(transform: Vector, rotation: Quaternion, scale: Vector) -> PyResult<String> {
    let mut type_registry = TypeRegistry::default();
    type_registry.register::<Quat>();

    let mut value = Transform::from_translation(transform.to_vec3());
    value.rotation = rotation.to_quat();
    value.scale = scale.to_vec3();

    let patch = DynamicStruct::default();
    value.apply(&patch);

    let serializer = ReflectSerializer::new(&value, &type_registry);
    let ron_string =
        ron::ser::to_string_pretty(&serializer, ron::ser::PrettyConfig::default()).unwrap();
    Ok(ron_string)
}

/// A Python module implemented in Rust.
#[pymodule]
fn bevy_adapter(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(transform_ron, m)?)?;
    Ok(())
}
