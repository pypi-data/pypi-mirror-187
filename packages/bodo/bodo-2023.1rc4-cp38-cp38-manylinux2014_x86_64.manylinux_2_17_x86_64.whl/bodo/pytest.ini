[pytest]
# https://dev.azure.com/bodo-inc/Bodo/_test/analytics?definitionId=5&contextType=build
# The default 350 sec timeout was determined by finding the upper bound for the majority of tests. 
# There exist other unit tests which take up more time than 350 sec,
# so @pytest.mark.timeout decorators are used to override the default timeout.
timeout = 350
markers =
    slow: marks tests as slow (deselect with '-m "not slow"')
    weekly: marks tests only on azure weekly (runs every Wednesday)
    smoke: marks tests as basic smoke tests to run on every PR (select with '-m "smoke"')
    s3: s3 related tests
    # Used for Azure NP=2
    hdfs: hdfs related tests
    bodo_1of10: part 1/10 of the test suite
    bodo_2of10: part 2/10 of the test suite
    bodo_3of10: part 3/10 of the test suite
    bodo_4of10: part 4/10 of the test suite
    bodo_5of10: part 5/10 of the test suite
    bodo_6of10: part 6/10 of the test suite
    bodo_7of10: part 7/10 of the test suite
    bodo_8of10: part 8/10 of the test suite
    bodo_9of10: part 9/10 of the test suite
    bodo_10of10: part 10/10 of the test suite
    bodosql_1of7: part 1/7 of the test suite
    bodosql_2of7: part 2/7 of the test suite
    bodosql_3of7: part 3/7 of the test suite
    bodosql_4of7: part 4/7 of the test suite
    bodosql_5of7: part 5/7 of the test suite
    bodosql_6of7: part 6/7 of the test suite
    bodosql_7of7: part 7/7 of the test suite
    # Used for Azure NP=1
    bodo_1of6: part 1/6 of the test suite
    bodo_2of6: part 2/6 of the test suite
    bodo_3of6: part 3/6 of the test suite
    bodo_4of6: part 4/6 of the test suite
    bodo_5of6: part 5/6 of the test suite
    bodo_6of6: part 6/6 of the test suite
    bodosql_1of4: part 1/4 of the test suite
    bodosql_2of4: part 2/4 of the test suite
    bodosql_3of4: part 3/4 of the test suite
    bodosql_4of4: part 4/4 of the test suite
    single_mod: single python module from which to run tests
    iceberg: iceberg related tests

    # Groups 
    1: AWS group 1
    2: AWS group 2
    3: AWS group 3
    4: AWS group 4
    5: AWS group 5
    6: AWS group 6
    7: AWS group 7
    8: AWS group 8
    9: AWS group 9
addopts = -p no:faulthandler
