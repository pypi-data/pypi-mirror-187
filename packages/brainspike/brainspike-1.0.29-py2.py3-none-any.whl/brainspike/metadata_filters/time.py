"""
times.py 

Modules for filtering recordings 
times in metadata. 

"""

import numpy as np

import pandas as pd

from datetime import date

###################################################################################################
###################################################################################################

def times(time_range = None): 
    """ return a range of times for metadata filtering """
    
    if time_range is not None: 
        if isinstance(time_range, list): 
            if len(time_range) == 2: 
                times = pd.date_range(start = time_range[0], end = time_range[1], freq="s")
    
    else: 
        raise KeyError("no list of times passed ...")

    return times.strftime("%H:%M:%S")

