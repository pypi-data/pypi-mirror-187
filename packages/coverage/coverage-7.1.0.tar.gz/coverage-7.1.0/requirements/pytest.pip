#
# This file is autogenerated by pip-compile with Python 3.7
# by the following command:
#
#    make upgrade
#
attrs==22.2.0 \
    --hash=sha256:29e95c7f6778868dbd49170f98f8818f78f3dc5e0e37c0b1f474e3561b240836 \
    --hash=sha256:c9227bfc2f01993c03f68db37d1d15c9690188323c067c641f1a35ca58185f99
    # via
    #   hypothesis
    #   pytest
colorama==0.4.6 \
    --hash=sha256:08695f5cb7ed6e0531a20572697297273c47b8cae5a63ffc6d6ed5c201be6e44 \
    --hash=sha256:4f1d9991f5acc0ca119f9d443620b77f9d6b33703e51011c16baf57afb285fc6
    # via -r requirements/pytest.in
exceptiongroup==1.1.0 \
    --hash=sha256:327cbda3da756e2de031a3107b81ab7b3770a602c4d16ca618298c526f4bec1e \
    --hash=sha256:bcb67d800a4497e1b404c2dd44fca47d3b7a5e5433dbab67f96c1a685cdfdf23
    # via
    #   hypothesis
    #   pytest
execnet==1.9.0 \
    --hash=sha256:8f694f3ba9cc92cab508b152dcfe322153975c29bda272e2fd7f3f00f36e47c5 \
    --hash=sha256:a295f7cc774947aac58dde7fdc85f4aa00c42adf5d8f5468fc630c1acf30a142
    # via pytest-xdist
flaky==3.7.0 \
    --hash=sha256:3ad100780721a1911f57a165809b7ea265a7863305acb66708220820caf8aa0d \
    --hash=sha256:d6eda73cab5ae7364504b7c44670f70abed9e75f77dd116352f662817592ec9c
    # via -r requirements/pytest.in
hypothesis==6.62.1 \
    --hash=sha256:7d1e2f9871e6509662da317adf9b4aabd6b38280fb6c7930aa4f574d2ed25150 \
    --hash=sha256:d00a4a9c54b0b8b4570fe1abe42395807a973b4a507e6718309800e6f84e160d
    # via -r requirements/pytest.in
importlib-metadata==6.0.0 \
    --hash=sha256:7efb448ec9a5e313a57655d35aa54cd3e01b7e1fbcf72dce1bf06119420f5bad \
    --hash=sha256:e354bedeb60efa6affdcc8ae121b73544a7aa74156d047311948f6d711cd378d
    # via
    #   pluggy
    #   pytest
iniconfig==2.0.0 \
    --hash=sha256:2d91e135bf72d31a410b17c16da610a82cb55f6b0477d1a902134b24a455b8b3 \
    --hash=sha256:b6a85871a79d2e3b22d2d1b94ac2824226a63c6b741c88f7ae975f18b6778374
    # via pytest
packaging==23.0 \
    --hash=sha256:714ac14496c3e68c99c29b00845f7a2b85f3bb6f1078fd9f72fd20f0570002b2 \
    --hash=sha256:b6ad297f8907de0fa2fe1ccbd26fdaf387f5f47c7275fedf8cce89f99446cf97
    # via pytest
pluggy==1.0.0 \
    --hash=sha256:4224373bacce55f955a878bf9cfa763c1e360858e330072059e10bad68531159 \
    --hash=sha256:74134bbf457f031a36d68416e1509f34bd5ccc019f0bcc952c7b909d06b37bd3
    # via pytest
pytest==7.2.1 \
    --hash=sha256:c7c6ca206e93355074ae32f7403e8ea12163b1163c976fee7d4d84027c162be5 \
    --hash=sha256:d45e0952f3727241918b8fd0f376f5ff6b301cc0777c6f9a556935c92d8a7d42
    # via
    #   -r requirements/pytest.in
    #   pytest-xdist
pytest-xdist==3.1.0 \
    --hash=sha256:40fdb8f3544921c5dfcd486ac080ce22870e71d82ced6d2e78fa97c2addd480c \
    --hash=sha256:70a76f191d8a1d2d6be69fc440cdf85f3e4c03c08b520fd5dc5d338d6cf07d89
    # via -r requirements/pytest.in
sortedcontainers==2.4.0 \
    --hash=sha256:25caa5a06cc30b6b83d11423433f65d1f9d76c4c6a0c90e3379eaa43b9bfdb88 \
    --hash=sha256:a163dcaede0f1c021485e957a39245190e74249897e2ae4b2aa38595db237ee0
    # via hypothesis
tomli==2.0.1 \
    --hash=sha256:939de3e7a6161af0c887ef91b7d41a53e7c5a1ca976325f429cb46ea9bc30ecc \
    --hash=sha256:de526c12914f0c550d15924c62d72abc48d6fe7364aa87328337a31007fe8a4f
    # via pytest
typing-extensions==4.4.0 \
    --hash=sha256:1511434bb92bf8dd198c12b1cc812e800d4181cfcb867674e0f8279cc93087aa \
    --hash=sha256:16fa4864408f655d35ec496218b85f79b3437c829e93320c7c9215ccfd92489e
    # via importlib-metadata
zipp==3.11.0 \
    --hash=sha256:83a28fcb75844b5c0cdaf5aa4003c2d728c77e05f5aeabe8e95e56727005fbaa \
    --hash=sha256:a7a22e05929290a67401440b39690ae6563279bced5f314609d9d03798f56766
    # via importlib-metadata
