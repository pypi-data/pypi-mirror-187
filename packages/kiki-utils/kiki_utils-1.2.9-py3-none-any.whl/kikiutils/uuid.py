from uuid import uuid1

from .file import read_file, save_file


# UUID

def get_uuid(save_path: str = './uuid.uuid'):
    if now_uuid := read_file(save_path):
        return now_uuid.decode()

    now_uuid = str(uuid1())
    save_file(save_path, now_uuid)
    return now_uuid
