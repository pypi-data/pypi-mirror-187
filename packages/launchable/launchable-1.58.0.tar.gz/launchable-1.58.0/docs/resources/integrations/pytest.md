---
description: This page outlines how the Launchable interfaces with pytest.
---

# pytest

{% hint style="info" %}
This is a reference page. See [Getting started](../../sending-data-to-launchable/using-the-launchable-cli/getting-started/), [Sending data to Launchable](../../sending-data-to-launchable/), and [Subsetting your test runs](../../features/predictive-test-selection/) for more comprehensive usage guidelines.
{% endhint %}

Launchable interfaces with pytest via

1. the Launchable pytest plugin
2. the Launchable CLI

## Launchable pytest plugin

We offer a new way to integrate Launchable, a native pytest plugin.

### Installing the plugin

The Launchable pytest plugin is a Python3 package that you can install from [PyPI](https://pypi.org/project/pytest-launchable/).

{% hint style="warning" %}
The plugin requires **Python 3.7+**, **Pytest 4.2.0+**, _and_ **Java 8+**.
{% endhint %}

If you use Pipenv, you can install the plugin into your repository:

```bash
pipenv install --dev pytest-launchable
```

Or, you can install the CLI in your CI pipeline by adding this to the part of your CI script where you install dependencies:

```bash
pip3 install pytest-launchable
```

You don't need to install Lanchable CLI separately because the plugin automatically installs the CLI and uses it internally.

### Setting your API key

First, create an API key for your workspace at [app.launchableinc.com](https://app.launchableinc.com). This authentication token allows the pytest plugin to talk to Launchable.

Then, make this API key available as the `LAUNCHABLE_TOKEN` environment variable in your CI process. How you do this depends on your CI system:

| CI system              | Docs                                                                                                                                                                                                 |
| ---------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Azure DevOps Pipelines | [Set secret variables](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops\&tabs=yaml%2Cbatch#secret-variables)                                              |
| Bitbucket Pipelines    | [Variables and secrets](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/)                                                                                                   |
| CircleCI               | [Using Environment Variables](https://circleci.com/docs/2.0/env-vars/)                                                                                                                               |
| GitHub Actions         | [How to configure a secret](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)                                                                                     |
| GitLab CI              | [GitLab CI/CD environment variables](https://docs.gitlab.com/ee/ci/variables/)                                                                                                                       |
| GoCD                   | [Setting variables on an environment](https://docs.gocd.org/current/faq/dev\_use\_current\_revision\_in\_build.html#setting-variables-on-an-environment)                                             |
| Jenkins                | <p><a href="https://docs.cloudbees.com/docs/cloudbees-ci/latest/cloud-secure-guide/injecting-secrets">Injecting secrets into builds</a></p><p>(Create a global "secret text" to use in your job)</p> |
| Travis CI              | [Environment Variables](https://docs.travis-ci.com/user/environment-variables/)                                                                                                                      |

### Generate a config file

`launchable-config` is a command-line tool to generate and validate configuration files. The Launchable pytest plugin uses this config.

First, generate a new config file:

```bash
# via pipenv
pipenv run launchable-config --create

# via pip
launchable-config --create
```

This generates a template `.launchable.d/config.yml` file in the current directory that looks like this:

```yml
# Launchable test session configuration file
# See https://docs.launchableinc.com/resources/cli-reference for detailed usage of these options
#  
schema-version: 1.0
build-name: commit_hash
record-build:
  # Put your git repository location here
  source: .
  max_days: 30
record-session:
subset:
  # mode is subset, subset-and-rest, or record-only
  mode: subset
  # you must specify one of target/confidence/time
  # examples:
  #   target: 30%  # Create a variable time-based subset of the given percentage. (0%-100%)
  #   confidence: 30%  # Create a confidence-based subset of the given percentage. (0%-100%)
  #   time: 30m  # Create a fixed time-based subset. Select the best set of tests that run within the given time bound. (e.g. 10m for 10 minutes, 2h30m for 2.5 hours, 1w3d for 7+3=10 days. )
  target: 30%
record-tests:
  # The test results are placed here in JUnit XML format
  result_dir: launchable-test-result
```

You can then edit the config file per the directions below.

### Recording test results (pytest plugin)

#### Update your config file

In `.launchable.d/config.yml`:

1. Check that the `source` option in the `record-build` section points to your Git repository (the default is `.`, the current directory).
2. Check that the `mode` option in the `subset` section is set to `record-only`

#### Verify your config file

Verify the contents of the `.launchable.d/config.yml` file:

```bash
# via pipenv
pipenv run launchable-config --verify

# via pip
launchable-config --verify
```

If any problems are reported, edit the file accordingly.

#### Use the plugin with pytest

Then, just add an `--launchable` option to the pytest command. It is very easy:

```bash
pytest --launchable <your-pytest-project>
```

If the configuration file is not in the current directory, use the `--launchable-conf-path` option:

```bash
pytest --launchable --launchable-conf-path <path-to-launchable-configuration-file> <your-pytest-project>
```

This will:

1. Create a build in your Launchable workspace
2. Run your tests
3. Submit your test reports to Launchable
4. Leave XML reports in the `launchable-test-result` by default

### Requesting and running a subset of tests

#### Update your config file

In `.launchable.d/config.yml`:

1. Check that the `source` option in the `record-build` section points to your Git repository (the default is `.`, the current directory).
2. Check that the `mode` option in the `subset` section is set to `subset` or `subset_and_rest` [based on your needs](../../features/predictive-test-selection/#training-wheels-mode-with-the-rest-option)
3. Check that one of the three [optimization target options](../../features/predictive-test-selection/#choosing-an-optimization-target) are set (`target`, `confidence`, or `time`)

#### Verify your config file

Verify the contents of the `.launchable.d/config.yml` file:

```bash
# via pipenv
pipenv run launchable-config --verify

# via pip
launchable-config --verify
```

If any problems are reported, edit the file accordingly.

#### Use the plugin with pytest

Then, just add an `--launchable` option to the pytest command. It is very easy:

```bash
pytest --launchable <your-pytest-project>
```

If the configuration file is not in the current directory, use the `--launchable-conf-path` option:

```bash
pytest --launchable --launchable-conf-path <path-to-launchable-configuration-file> <your-pytest-project>
```

This will:

1. Create a build in your Launchable workspace
2. Request a subset of tests based on your optimization target
3. Run those tests (or run all the tests if `subset_and_rest` mode is chosen)
4. Submit your test reports to Launchable
5. Leave XML reports in the `launchable-test-result` by default

## Launchable CLI

See [recording-test-results-with-the-launchable-cli](../../sending-data-to-launchable/using-the-launchable-cli/recording-test-results-with-the-launchable-cli/ "mention") and [subsetting-with-the-launchable-cli](../../features/predictive-test-selection/requesting-and-running-a-subset-of-tests/subsetting-with-the-launchable-cli/ "mention") for more information.
