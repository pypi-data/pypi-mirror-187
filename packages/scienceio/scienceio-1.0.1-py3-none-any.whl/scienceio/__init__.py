import os

from .scienceio import HTTPError, ScienceIO, ScienceIOError  # noqa

__version__ = os.environ.get("SCIENCEIO_SDK_VERSION", "v0.0.1")
