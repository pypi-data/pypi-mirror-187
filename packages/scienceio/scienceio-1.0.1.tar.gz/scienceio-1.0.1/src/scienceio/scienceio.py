import aiohttp
import asyncio
import configparser
import os
import requests
import time
import types

from enum import Enum
from pathlib import Path
from typing import Dict, Optional, Union

CONFIG = types.SimpleNamespace(
    API_URL="https://api.aws.science.io/v2",
    HELP_EMAIL="api_support@science.io",
    SETTINGS_DIR=Path.home() / ".scienceio",
    SETTINGS_FILE_PATH=Path.home() / ".scienceio" / "config",
    MAX_POLL_DURATION_SEC=300,
    POLL_SLEEP_DURATION_SEC=2,
)


class ScienceIOError(Exception):
    """Base class for all exceptions that are raised by the ScienceIO SDK."""


class HTTPError(ScienceIOError):
    """Raised when an HTTP error occurs when using the ScienceIO SDK."""

    def __init__(self, status_code: int, message: str):
        super().__init__(message)
        self.status_code = status_code


class TimeoutError(ScienceIOError):
    """Raised when a call to the ScienceIO API times out."""

    def __init__(self, msg):
        self.msg = msg


class ResponseFormat(str, Enum):
    JSON = "application/json"

    def format_response(self, response):
        """
        Defines response format

        Args:
            response: string with desired format
        """
        if self is ResponseFormat.JSON:
            return response.json()
        else:
            raise NotImplementedError("Unknown response format")

    def create_headers(self, api_id: str, api_secret: str) -> dict:
        """
        Create headers for API call

        Args:
            api_id: API Key, obtained from api.science.io
            api_secret: API secret, obtained from api.science.io

        Returns:
            Dict containing content type and API credentials
        """
        return {
            "Content-Type": self.value,
            "x-api-id": api_id,
            "x-api-secret": api_secret,
        }


class ScienceIO(object):
    def __init__(
        self,
        response_format: ResponseFormat = ResponseFormat.JSON,
        timeout: Optional[Union[int, float]] = 1200,
    ):
        """Initializer for the ScienceIO client. The client will attempt to
        configure itself by trying to read from environment variables, if set.
        If unable to, the config values will be read from the ScienceIO config
        file.

        Args:
            response_format:
                Format to use for responses (default: `ResponseFormat.JSON`).
            timeout:
                Amount of time to wait before timing out network calls, in seconds.
                If `None`, timeouts will be disabled. (default: 1200)
        """

        self.response_format = response_format
        self.timeout = timeout

        # Create a persistent session across requests.
        # https://docs.python-requests.org/en/master/user/advanced/
        self.session = requests.Session()
        self.session.params = {}

        # Lazy loading of configuration (no need to try and load the settings if user specifies
        # their own API ID and secret). Also, this prevents breakage when in envs with no
        # settings file, such as test environments or hosted Jupyter notebooks.
        config = None

        def get_config_value(key: str) -> str:
            nonlocal config
            if config is None:
                config = configparser.RawConfigParser()
                config.read(CONFIG.SETTINGS_FILE_PATH)

            return config["SETTINGS"][key].strip("\"'")

        # Handles config values from user arguments, config file, and user
        # input, in that order from most to least preferred.
        write_out_conf = False

        def get_value(initial: Optional[str], key: str, human_name: str) -> str:
            nonlocal write_out_conf
            if initial is None:
                try:
                    return get_config_value(key)
                except KeyError:
                    user_input = str(
                        input(f"Please provide your ScienceIO API key {human_name}: ")
                    )

                    # User input was collected, flag the config file for rewriting.
                    write_out_conf = True
                    return user_input

            return initial

        # API endpoints to use.
        self.api_url = os.environ.get("SCIENCEIO_API_URL", CONFIG.API_URL)
        self.structure_url = f"{self.api_url}/structure"

        # API key and secret (with extra handling for env vars, config file, and user prompts).
        self.api_id = get_value(os.environ.get("SCIENCEIO_KEY_ID"), "KEY_ID", "id")
        self.api_secret = get_value(
            os.environ.get("SCIENCEIO_KEY_SECRET"), "KEY_SECRET", "secret"
        )

        # Construct the headers.
        self.headers = self.response_format.create_headers(
            api_id=self.api_id,
            api_secret=self.api_secret,
        )

        # Write out the API key ID and secret if either or both of those values
        # needed user input.
        if write_out_conf:
            # Create a new `ConfigParser` to hold the configuration we want to
            # write to the config file.
            new_conf = configparser.ConfigParser()
            new_conf["SETTINGS"] = {
                "KEY_ID": self.api_id,
                "KEY_SECRET": self.api_secret,
            }

            # Create the config directory (and any parents) if it does not
            # already exist.
            CONFIG.SETTINGS_DIR.mkdir(parents=True, exist_ok=True)

            # Write out the new config.
            with open(CONFIG.SETTINGS_FILE_PATH, "w") as fp:
                new_conf.write(fp)

    def _construct_poll_url(self, request_id: str) -> str:
        """Helper method to construct a polling URL."""
        return f"{self.structure_url}/{request_id}"

    def annotate(self, *args, **kwargs) -> dict:
        return self.structure(*args, **kwargs)

    def structure(self, text: str) -> dict:
        """Annotates a block of text using the ScienceIO API.

        Args:
            text (str): The text to structure.

        Raises:
            ValueError: Raised if the input text is not a `str`.
            TimeoutError: Raised if the annotation request exceeds the timeout
                limit.

        Returns:
            dict: The annotation data, as a Python dictionary.
        """
        request_id = self.send_structure_request(text)

        poll_url = self._construct_poll_url(request_id)

        # Poll for the response.
        first_loop = True
        start_time = time.time()
        while first_loop or time.time() - start_time < CONFIG.MAX_POLL_DURATION_SEC:
            # Sleep for a bit if we're not on the first loop.
            if not first_loop:
                time.sleep(CONFIG.POLL_SLEEP_DURATION_SEC)

            first_loop = False

            inf_result = self._poll_attempt(poll_url)

            if inf_result is not None:
                return inf_result

            # At this point, the request is still processing, so try again.

        # Here, we've exhausted all of our retry attempts, so fail.
        raise TimeoutError("structure request timed out, try again later")

    def send_annotate_request(self, *args, **kwargs) -> str:
        return self.send_structure_request(*args, **kwargs)

    def send_structure_request(self, text: str) -> str:
        """Submits a request to structure a block of text using the ScienceIO API.

        Args:
            text (str): The text to structure.

        Returns:
            str: A request token that is used with `get_annotation_reponse` to
                check the status of the annotation request, and to fetch the
                results when ready.
        """
        raw_response = self.session.post(
            self.structure_url,
            json={"text": text},
            headers=self.headers,
        )

        response = _response_handler(raw_response, self.response_format)

        request_id = response["request_id"]

        return request_id

    def get_annotate_response(self, *args, **kwargs) -> Optional[Dict]:
        return self.get_structure_response(*args, **kwargs)

    def get_structure_response(self, request_id: str) -> Optional[Dict]:
        """Given a request token, returns annotations if the corresponding
        annotation request has completed successfully.

        Args:
            request_id (str): The request token returned from a prior call to
                `send_structure_request`.

        Returns:
            Optional[Dict]: The annotations if the request has completed
                successfully, or `None` if it is still processing.
        """
        # Construct the URL to poll.
        poll_url = self._construct_poll_url(request_id)
        return self._poll_attempt(poll_url)

    async def annotate_async(self, *args, **kwargs):
        return self.structure_async(*args, **kwargs)

    async def structure_async(self, text: str):
        """Asynchronously structures a block of text using the ScienceIO API.

        Args:
            text (str): The text to structure.

        Raises:
            TimeoutError: Raised if the annotation request exceeds the timeout
                limit.

        Returns:
            dict: The annotation data, as a Python dictionary.
        """
        request_id = self.send_structure_request(text)

        async with aiohttp.ClientSession() as session:
            async with session.post(
                self.structure_url, json={"text": text}, headers=self.headers
            ) as response_future:
                response = await _response_handler_async(response_future)

                request_id = response["request_id"]

            poll_url = self._construct_poll_url(request_id)

            # Poll for the response.
            first_loop = True
            start_time = time.time()
            while (
                first_loop or (time.time() - start_time) < CONFIG.MAX_POLL_DURATION_SEC
            ):
                # Sleep for a bit if we're not on the first loop.
                if not first_loop:
                    await asyncio.sleep(CONFIG.POLL_SLEEP_DURATION_SEC)

                first_loop = False

                async with session.get(
                    poll_url, headers=self.headers
                ) as response_future:
                    response = await _response_handler_async(response_future)

                inf_result = _poll_payload_handler(response)

                if inf_result is not None:
                    return inf_result

                # At this point, the request is still processing, so try again.

            # Here, we've exhausted all of our retry attempts, so fail.
            raise TimeoutError("structure request timed out, try again later")

    def _poll_attempt(self, poll_url: str) -> Optional[Dict]:
        """Helper method to make a single poll attempt, given a constructed
        poll URL with a request id."""
        raw_response = self.session.get(
            poll_url,
            headers=self.headers,
        )

        raw_response.raise_for_status()

        response = self.response_format.format_response(raw_response)

        return _poll_payload_handler(response)


def _response_handler(
    raw_response: requests.Response, response_format: ResponseFormat
) -> Dict:
    status_code = raw_response.status_code
    response = response_format.format_response(raw_response)

    if 400 <= status_code <= 599:
        details = response["detail"]
        raise HTTPError(
            status_code=status_code,
            message=f"{', '.join(details['errors']['text'])}",
        )

    return response


async def _response_handler_async(response_future: aiohttp.ClientResponse) -> Dict:
    status_code = response_future.status
    response = await response_future.json()

    if 400 <= status_code <= 599:
        details = response["detail"]
        raise HTTPError(
            status_code=status_code,
            message=f"{', '.join(details['errors']['text'])}",
        )

    return response


def _poll_payload_handler(payload: Dict) -> Optional[Dict]:
    status = payload["inference_status"]
    if status == "ERRORED" or status == "EXPIRED":
        raise ScienceIOError(payload["message"])
    elif status == "COMPLETED":
        return payload["inference_result"]
    elif status == "SUBMITTED":
        return None
    else:
        raise ScienceIOError("unknown status")
