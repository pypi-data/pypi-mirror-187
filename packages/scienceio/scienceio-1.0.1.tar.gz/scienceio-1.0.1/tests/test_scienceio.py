import pytest
import time
import uuid

from src.scienceio import HTTPError, ScienceIO

QUERY_TEXT = (
    "The COVID-19 pandemic has shown a markedly low proportion of cases among "
    "children 1-4. Age disparities in observed cases could be explained by "
    "children having lower susceptibility to infection, lower propensity to "
    "show clinical symptoms or both."
)


@pytest.fixture(scope="session", autouse=True)
def scio():
    """
    Init ScienceIO client, using .scio/config
    :return: our client
    """
    yield ScienceIO()


def test_input_is_string(scio):
    """
    Test providing something other than a string returns an error
    """
    # Invalid types.
    with pytest.raises(HTTPError, match="none is not an allowed value"):
        scio.annotate(text=None)

    with pytest.raises(HTTPError, match="str type expected"):
        scio.annotate(text={})

    with pytest.raises(HTTPError, match="str type expected"):
        scio.annotate(text={"nested": "dict"})

    # TODO: Uncomment when this is working in NRT.
    # with pytest.raises(HTTPError, match="str type expected"):
    #     scio.annotate(text=123)

    # Empty strings should be rejected.
    with pytest.raises(HTTPError, match="ensure this value has at least 1 characters"):
        scio.annotate(text="")

    scio.structure(text="normal strings should work just fine")


def test_structure(scio, snapshot):
    """
    Test structure request with a small text snippet
    """

    response = scio.structure(text=QUERY_TEXT)

    assert isinstance(response, dict)

    # Ensure floating-point keys exist in spans, but scrub them out since they
    # may vary in value from run to run slightly and are not always comparable.
    for span in response["spans"]:
        del span["score_id"]
        del span["score_type"]

    assert response == snapshot


def test_async_annotation(scio):
    """
    Test async structure request with a small text snippet.
    """
    request_id = scio.send_structure_request(text=QUERY_TEXT)

    # The request id would be a valid UUID string.
    assert isinstance(request_id, str)
    uuid.UUID(request_id)

    time.sleep(10)

    response = scio.get_structure_response(request_id)

    assert isinstance(response, dict)


def test_async_def_annotation(scio, snapshot):
    """
    Test async structure request with a small text snippet.
    """
    import asyncio

    response = asyncio.run(scio.structure_async(text=QUERY_TEXT))

    assert isinstance(response, dict)

    # Ensure floating-point keys exist in spans, but scrub them out since they
    # may vary in value from run to run slightly and are not always comparable.
    for span in response["spans"]:
        del span["score_id"]
        del span["score_type"]

    assert response == snapshot
