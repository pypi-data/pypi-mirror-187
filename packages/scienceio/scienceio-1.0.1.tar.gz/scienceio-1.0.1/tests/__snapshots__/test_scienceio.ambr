# name: test_structure
  dict({
    'spans': list([
      dict({
        'concept_id': 'UMLS:C5203670',
        'concept_name': 'COVID-19',
        'concept_type': 'Medical Conditions',
        'pos_end': 12,
        'pos_start': 4,
        'text': 'COVID-19',
      }),
      dict({
        'concept_id': 'UMLS:C1615608',
        'concept_name': 'Pandemics',
        'concept_type': 'Anatomy & Physiology',
        'pos_end': 21,
        'pos_start': 13,
        'text': 'pandemic',
      }),
      dict({
        'concept_id': 'UMLS:C0008059',
        'concept_name': 'Child',
        'concept_type': 'Context',
        'pos_end': 81,
        'pos_start': 73,
        'text': 'children',
      }),
      dict({
        'concept_id': 'UMLS:C0001779',
        'concept_name': 'Age',
        'concept_type': 'Anatomy & Physiology',
        'pos_end': 90,
        'pos_start': 87,
        'text': 'Age',
      }),
      dict({
        'concept_id': 'UMLS:C0008059',
        'concept_name': 'Child',
        'concept_type': 'Context',
        'pos_end': 151,
        'pos_start': 143,
        'text': 'children',
      }),
      dict({
        'concept_id': 'UMLS:C0009450',
        'concept_name': 'Communicable Diseases',
        'concept_type': 'Medical Conditions',
        'pos_end': 192,
        'pos_start': 183,
        'text': 'infection',
      }),
    ]),
    'text': 'The COVID-19 pandemic has shown a markedly low proportion of cases among children 1-4. Age disparities in observed cases could be explained by children having lower susceptibility to infection, lower propensity to show clinical symptoms or both.',
  })
# ---
# name: test_async_def_annotation
  dict({
    'spans': list([
      dict({
        'concept_id': 'UMLS:C5203670',
        'concept_name': 'COVID-19',
        'concept_type': 'Medical Conditions',
        'pos_end': 12,
        'pos_start': 4,
        'text': 'COVID-19',
      }),
      dict({
        'concept_id': 'UMLS:C1615608',
        'concept_name': 'Pandemics',
        'concept_type': 'Anatomy & Physiology',
        'pos_end': 21,
        'pos_start': 13,
        'text': 'pandemic',
      }),
      dict({
        'concept_id': 'UMLS:C0008059',
        'concept_name': 'Child',
        'concept_type': 'Context',
        'pos_end': 81,
        'pos_start': 73,
        'text': 'children',
      }),
      dict({
        'concept_id': 'UMLS:C0001779',
        'concept_name': 'Age',
        'concept_type': 'Anatomy & Physiology',
        'pos_end': 90,
        'pos_start': 87,
        'text': 'Age',
      }),
      dict({
        'concept_id': 'UMLS:C0008059',
        'concept_name': 'Child',
        'concept_type': 'Context',
        'pos_end': 151,
        'pos_start': 143,
        'text': 'children',
      }),
      dict({
        'concept_id': 'UMLS:C0009450',
        'concept_name': 'Communicable Diseases',
        'concept_type': 'Medical Conditions',
        'pos_end': 192,
        'pos_start': 183,
        'text': 'infection',
      }),
    ]),
    'text': 'The COVID-19 pandemic has shown a markedly low proportion of cases among children 1-4. Age disparities in observed cases could be explained by children having lower susceptibility to infection, lower propensity to show clinical symptoms or both.',
  })
# ---
